java-frontend/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java

```diff
@@ -502,6 +512,59 @@ private static ExpressionTree cleanupCondition(ExpressionTree condition) {
     return cleanedUpCondition;
   }
 
+  private void handleSwitch(CFG.Block programPosition, List<CaseGroupTree> caseGroups) {
+    ProgramState.Pop poppedSwitchValue = programState.unstackValue(1);
+    ProgramState.SymbolicValueSymbol switchValue = poppedSwitchValue.valuesAndSymbols.get(0);
+    ProgramState state = poppedSwitchValue.state;
+
+    Map<CaseGroupTree, List<ProgramState.SymbolicValueSymbol>> caseValues = new HashMap<>();
+    for (CaseGroupTree caseGroup : caseGroups) {
+      int numberOfCaseValues = caseGroup.labels()
+        .stream()
+        .map(CaseLabelTree::expressions)
+        .mapToInt(List::size)
+        .sum();
+      ProgramState.Pop poppedCaseValues = state.unstackValue(numberOfCaseValues);
+      state = poppedCaseValues.state;
+      caseValues.put(caseGroup, poppedCaseValues.valuesAndSymbols);
+    }
+
+    ProgramState elseState = state;
+    // The block that will be taken when all case-conditions are false. This will either be the default-block or, if no
+    // default block exists, the block after the switch statement.
+    CFG.Block elseBlock = null;
+    for (CFG.Block successor : programPosition.successors()) {
+      CaseGroupTree caseGroup = successor.caseGroup();
+      if (caseGroup == null || !caseValues.containsKey(caseGroup)) {
+        assert(elseBlock == null);
+        elseBlock = successor;
+        continue;
+      }
+
+      for (ProgramState.SymbolicValueSymbol caseValue : caseValues.get(caseGroup)) {
+        SymbolicValue equality = constraintManager.createEquality(switchValue, caseValue);
+        ProgramState ps = setConstraint(state, equality, BooleanConstraint.TRUE);
+        enqueue(new ProgramPoint(successor), ps, node.exitPath);
+        elseState = setConstraint(elseState, equality, BooleanConstraint.FALSE);
+      }
+      if (successor.isDefaultBlock()) {
+        assert(elseBlock == null);
+        elseBlock = successor;
+      }
+    }
+    assert(elseBlock != null);
+    enqueue(new ProgramPoint(elseBlock), elseState, node.exitPath);
+  }
+
+  private static ProgramState setConstraint(ProgramState state, SymbolicValue condition, BooleanConstraint constraint) {
```

Comment:
I'm a bit unsure about this method because I'm generally a bit unsure about what SymbolicValue.setConstraint does (as opposed to ProgramState.addConstraint). From the code it looks like it adds the constraint unless it's already there, but that doesn't explain why it failed to properly add the constraints when I used `ProgramState.addConstraint` instead of this method.